File: leetcode/2024_6_june/c++/11_relative_sort_array.cpp
// TIME COMPLEXITY
// O(n * logn)

// SPACE COMPLEXITY
// O(n)

// CODE

class Solution {
public:
    vector<int> relativeSortArray(vector<int>& arr1, vector<int>& arr2) {
        map<int,int> mp;
        for(int &i : arr1){
            mp[i]++;
        }

        int index = 0;

        for(int &i : arr2){
            while(mp[i]-- > 0){
                arr1[index++] = i;
            }
        }

        for(auto &i : mp){
            int freq = i.second;
            while(freq > 0){
                arr1[index++] = i.first;
                freq--;
            }
        }

        return arr1;
    }
};

File: leetcode/2024_6_june/python/11_relative_sort_array.py
# CODE

from typing import List
class Solution:
    def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
        maxiboi = max(arr1)
        count_sort = [0] * (maxiboi + 1)
        for val in arr1:
            count_sort[val] += 1
        res = []
        for val in arr2:
            while count_sort[val] > 0:
                res.append(val)
                count_sort[val] -= 1
        for val in range(maxiboi + 1):
            while count_sort[val] > 0:
                res.append(val)
                count_sort[val] -= 1
        return res